x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java

```diff
@@ -175,6 +176,18 @@ static ResolvedIndices resolveIndicesAndAliasesWithoutWildcards(String action, I
             }
             localIndices.add(IndexNameExpressionResolver.resolveDateMathExpression(name));
         }
+        /// MP TODO: best way to handle this?
+        if (indicesRequest instanceof IndicesRequest.SingleIndexNoWildcards single && single.allowsRemoteIndices()) {
+            assert indices.length == 1;
+
+            // ensure that `index` has a remote name and not a date math expression which includes ':' symbol
+            // since date math expression after evaluation should not contain ':' symbol
+            String indexExpression = IndexNameExpressionResolver.resolveDateMathExpression(indices[0]);
+            if (indexExpression.indexOf(RemoteClusterService.REMOTE_CLUSTER_INDEX_SEPARATOR) >= 0) {
+                return new ResolvedIndices(List.of(), List.of(indices[0]));
+            }
+        }
+
```

Comment:
I noticed that after the request is successfully authorized on the querying cluster and then forwarded to the fulfilling cluster, there it is again authorized with the same index expression containing again **the index prefixed with the remote cluster**. But that's different from how authz code is expecting things. Authz code expects that the request on the fulfilling cluster to not include the remote cluster name. If they do they are again authz as if they're again a remote index request (but it's a local index request now).

So I think you need to drop the remote cluster name prefix on index names when forwarding to a remote cluster.

Reply:
The code here I think almost works, BUT it uses an uncoventional way to separate remote from local indices, as well as evaluating datemaths for remote index names. I think these might cause subtle problems.

I've put together this diff, which instead reuses the same method to split local/remote indices form expressions that's used for regular indices requests:
```
diff --git a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java
index c8920d0f498..c886f2fde55 100644
--- a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java
+++ b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java
@@ -477,7 +477,7 @@ public class AuthorizationService {
                     resolvedIndicesListener.onResponse(resolvedIndices);
                     return;
                 }
-                final ResolvedIndices resolvedIndices = IndicesAndAliasesResolver.tryResolveWithoutWildcards(action, request);
+                final ResolvedIndices resolvedIndices = indicesAndAliasesResolver.tryResolveWithoutWildcards(action, request);
                 if (resolvedIndices != null) {
                     resolvedIndicesListener.onResponse(resolvedIndices);
                 } else {
@@ -870,8 +870,8 @@ public class AuthorizationService {
         }, listener::onFailure));
     }
 
-    private static String resolveIndexNameDateMath(BulkItemRequest bulkItemRequest) {
-        final ResolvedIndices resolvedIndices = IndicesAndAliasesResolver.resolveIndicesAndAliasesWithoutWildcards(
+    private String resolveIndexNameDateMath(BulkItemRequest bulkItemRequest) {
+        final ResolvedIndices resolvedIndices = indicesAndAliasesResolver.resolveIndicesAndAliasesWithoutWildcards(
             getAction(bulkItemRequest),
             bulkItemRequest.request()
         );
diff --git a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java
index 5fbcc9f5b34..bf1bf7b7d3c 100644
--- a/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java
+++ b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java
@@ -28,7 +28,6 @@ import org.elasticsearch.core.Nullable;
 import org.elasticsearch.index.Index;
 import org.elasticsearch.index.IndexNotFoundException;
 import org.elasticsearch.transport.RemoteClusterAware;
-import org.elasticsearch.transport.RemoteClusterService;
 import org.elasticsearch.transport.RemoteConnectionStrategy;
 import org.elasticsearch.transport.TransportRequest;
 import org.elasticsearch.xpack.core.security.authz.AuthorizationEngine;
@@ -47,7 +46,6 @@ import java.util.SortedMap;
 import java.util.concurrent.CopyOnWriteArraySet;
 import java.util.function.Predicate;
 import java.util.function.Supplier;
-import java.util.stream.Stream;
 
 import static org.elasticsearch.xpack.core.security.authz.IndicesAndAliasesResolverField.NO_INDEX_PLACEHOLDER;
 
@@ -121,7 +119,7 @@ class IndicesAndAliasesResolver {
      * @return The {@link ResolvedIndices} or null if wildcard expansion must be performed.
      */
     @Nullable
-    static ResolvedIndices tryResolveWithoutWildcards(String action, TransportRequest transportRequest) {
+    ResolvedIndices tryResolveWithoutWildcards(String action, TransportRequest transportRequest) {
         // We only take care of IndicesRequest
         if (false == transportRequest instanceof IndicesRequest) {
             return null;
@@ -146,7 +144,7 @@ class IndicesAndAliasesResolver {
         return false;
     }
 
-    static ResolvedIndices resolveIndicesAndAliasesWithoutWildcards(String action, IndicesRequest indicesRequest) {
+    ResolvedIndices resolveIndicesAndAliasesWithoutWildcards(String action, IndicesRequest indicesRequest) {
         assert false == requiresWildcardExpansion(indicesRequest) : "request must not require wildcard expansion";
         final String[] indices = indicesRequest.indices();
         if (indices == null || indices.length == 0) {
@@ -163,32 +161,28 @@ class IndicesAndAliasesResolver {
             );
         }
 
+        final ResolvedIndices split;
+        if (indicesRequest instanceof IndicesRequest.SingleIndexNoWildcards single && single.allowsRemoteIndices()) {
+            split = remoteClusterResolver.splitLocalAndRemoteIndexNames(indicesRequest.indices());
+        } else {
+            split = new ResolvedIndices(Arrays.asList(indicesRequest.indices()), List.of());
+        }
+
         // NOTE: shard level requests do support wildcards (as they hold the original indices options) but don't support
         // replacing their indices.
         // That is fine though because they never contain wildcards, as they get replaced as part of the authorization of their
         // corresponding parent request on the coordinating node. Hence wildcards don't need to get replaced nor exploded for
         // shard level requests.
-        final List<String> localIndices = new ArrayList<>(indices.length);
-        for (String name : indices) {
+        final List<String> localIndices = new ArrayList<>(split.getLocal().size());
+        for (String localName : split.getLocal()) {
             // TODO: Shard level requests have wildcard expanded already and do not need go through this check
-            if (Regex.isSimpleMatchPattern(name)) {
-                throwOnUnexpectedWildcards(action, indices);
-            }
-            localIndices.add(IndexNameExpressionResolver.resolveDateMathExpression(name));
-        }
-        /// MP TODO: best way to handle this?
-        if (indicesRequest instanceof IndicesRequest.SingleIndexNoWildcards single && single.allowsRemoteIndices()) {
-            assert indices.length == 1;
-
-            // ensure that `index` has a remote name and not a date math expression which includes ':' symbol
-            // since date math expression after evaluation should not contain ':' symbol
-            String indexExpression = IndexNameExpressionResolver.resolveDateMathExpression(indices[0]);
-            if (indexExpression.indexOf(RemoteClusterService.REMOTE_CLUSTER_INDEX_SEPARATOR) >= 0) {
-                return new ResolvedIndices(List.of(), List.of(indices[0]));
+            if (Regex.isSimpleMatchPattern(localName)) {
+                throwOnUnexpectedWildcards(action, split.getLocal());
             }
+            localIndices.add(IndexNameExpressionResolver.resolveDateMathExpression(localName));
         }
 
-        return new ResolvedIndices(localIndices, List.of());
+        return new ResolvedIndices(localIndices, split.getRemote());
     }
 
     /**
@@ -209,8 +203,8 @@ class IndicesAndAliasesResolver {
         return split;
     }
 
-    private static void throwOnUnexpectedWildcards(String action, String[] indices) {
-        final List<String> wildcards = Stream.of(indices).filter(Regex::isSimpleMatchPattern).toList();
+    private static void throwOnUnexpectedWildcards(String action, List<String> indices) {
+        final List<String> wildcards = indices.stream().filter(Regex::isSimpleMatchPattern).toList();
         assert wildcards.isEmpty() == false : "we already know that there's at least one wildcard in the indices";
         throw new IllegalArgumentException(
             "the action "
```
I think this is a safe localized change to the authz logic to support the `IndicesRequest.SingleIndexNoWildcards` kind of requests.

Reply:
@albertzaharovits 

> I noticed that after the request is successfully authorized on the querying cluster and then forwarded to the fulfilling cluster, there it is again authorized with the same index expression containing again the index prefixed with the remote cluster.

Good catch. I had logic to strip it off on the remote cluster, but you're right, it is better to strip it off before sending it. I've fixed that in the [latest commit](https://github.com/elastic/elasticsearch/pull/105360/commits/93d6ff6f3d59ed4bf57c47a33aa744ca83983a33). 

> I've put together this diff

Thank you very much! I applied that diff. With the above change along with your patch, now the IT test behaves as I expect. Before in one place it was returning a 404 (index not found) where it should return a 403. Now it behaves correctly, I think.